gufe/serialization/msgpack.py

```diff
@@ -0,0 +1,138 @@
+import uuid
+from datetime import datetime
+from enum import IntEnum
+from pathlib import Path
+
+import msgpack
+import numpy as np
+import pint
+from openff.units import DEFAULT_UNIT_REGISTRY
+
+import gufe
+from gufe.settings import SettingsBaseModel
+
+
+# The integer values below represent the leading bytes of the extension format.
+# When adding a new type, manually assign the next available number (do not use auto()).
+# If you need to remove a type, avoid reusing its value in the future (preferably comment out the line).
+class MPEXT(IntEnum):
+    LARGEINT = 0
+    PATH = 1
+    UNIT = 2
+    NDARRAY = 3
+    NPGENERIC = 4
+    SETTINGS = 5
+    UUID = 6
+    DATETIME = 7
+
+
+def pack_largeint(large_int: int) -> bytes:
+    return msgpack.packb(str(large_int))
+
+
+def pack_default(obj) -> msgpack.ExtType:
+    """For non-standard datatypes, create an ExtType containing packed
+    data."""
+    match obj:
+        case datetime():
+            dt_payload: bytes = msgpack.packb(obj.isoformat())
+            return msgpack.ExtType(MPEXT.DATETIME, dt_payload)
+        case SettingsBaseModel():
+            settings_data = {field: getattr(obj, field) for field in obj.__fields__}
+            settings_data.update({"__class__": obj.__class__.__qualname__, "__module__": obj.__class__.__module__})
+            settings_payload: bytes = msgpack.packb(settings_data, default=pack_default)
+            return msgpack.ExtType(MPEXT.SETTINGS, settings_payload)
+        case pint.registry.UnitRegistry.Quantity():
+            unit_data: list = [obj.m, str(obj.u), "openff_units"]
+            unit_payload: bytes = msgpack.packb(unit_data, default=pack_default)
+            return msgpack.ExtType(
+                MPEXT.UNIT,
+                unit_payload,
+            )
+        case Path():
+            return msgpack.ExtType(MPEXT.PATH, msgpack.packb(str(obj)))
+        case np.generic():
+            npg_payload: bytes = msgpack.packb([str(obj.dtype), obj.tobytes()], default=pack_default)
+            return msgpack.ExtType(MPEXT.NPGENERIC, npg_payload)
+        case np.ndarray():
+            npa_payload: bytes = msgpack.packb([str(obj.dtype), list(obj.shape), obj.tobytes()], default=pack_default)
+            return msgpack.ExtType(MPEXT.NDARRAY, npa_payload)
+        case int():
+            try:
+                return msgpack.packb(obj)
+            except OverflowError:
```

Comment:
Can you add a comment giving what conditions this happens for? As in, how big of an int does `msgpack` natively support?